### ISSUE TITLE
Add option for exiting after Typegen step

### ISSUE THREAD

**Issue by tgriesser** [prisma-labs/nexus#272]
One common question is how to generate the types without running the server. A simple option would be to add a config option to the schema:

```ts
/**
 * If set to true, or if NEXUS_EXIT env var is set to `true`,
 * we will call process.exit() after the types are emitted.
 */
exitAfterTypegen: boolean
```

And then we can tell folks to just install `ts-node` and run:

```sh
NEXUS_EXIT=true node -r ts-node/register/transpile-only path/to/schema-file.ts
```

**Comment by jasonkuhrt** [comment]
CC @nikolasburk this directly impacts our chat yesterday.

This also impacts my just shipped example #274.

**Comment by jasonkuhrt** [comment]
Thinking a bit more about the environment variables. We currently have `NEXUS_SHOULD_GENERATE_ARTIFACTS` to perform typegen smoothly in certain deploy contexts where `NODE_ENV=production`. Now we're introducing another env var, say as I suggested before, `NEXUS_EXIT_AFTER_TYPEGEN`. That means a heroku deployment build script might look like this now:

```json
{
  "scripts": {
    "build": "NEXUS_SHOULD_GENERATE_ARTIFACTS=true NEXUS_EXIT_AFTER_TYPEGEN=true ts-node --transpile-only src"
  }
}
```

This is just not great DX.

The solution might be an alternative (or third) to `NEXUS_EXIT_AFTER_TYPEGEN` that represents these two ideas which are very closely related (when would you want `NEXUS_SHOULD_GENERATE_ARTIFACTS` _but not_ `NEXUS_EXIT_AFTER_TYPEGEN`?). I would call this a domain specific env var whereas the other two are more just projections from the code config.

```json
{
  "scripts": {
    "build": "NEXUS_TYPEGEN=true ts-node --transpile-only src"
  }
}
```

But we still have a relatively high bar here requiring 1) use `ts-node` 2) use `--transpile-only` flag. I think these are arguably internal details the user shouldn't have to think about. So perhaps the solution is not about (or just about) a domain specific env var but rather a higher level cli command:

```
$ nexus typegen
```

This abstraction remove _three/four_ things the user would otherwise have to think about:

1. `ts-node`
2. `--transpile-only`
3. `NEXUS_TYPEGEN` or `NEXUS_EXIT_AFTER_TYPEGEN` + `NEXUS_SHOULD_GENERATE_ARTIFACTS`

Finally, we already know that we want to start getting into nexus CLI features. This finally seems like a good time to embark on that journey.

Given a CLI, I think then that _our first step_ should not introduce domain specific env vars and instead leave it a systems pattern thing, env vars transparently reflecting code config options. The design principal here is keeping an abstract configurable core and letting the "shell" (cli) provide opinionated recipes against it.

That said I think sooner or later we'll refactor the core config where it becomes clear certain basics are just wrong, e.g. as stated before:

> when would you want `NEXUS_SHOULD_GENERATE_ARTIFACTS` _but not_ `NEXUS_EXIT_AFTER_TYPEGEN`?

**Comment by jasonkuhrt** [comment]
Had a nice conversation with @timsuchanek and he raised an interesting idea of exploring the potential of performing Nexus typegen with with [TypeScript Plugins](https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin). It seems there could be some nice benefits to this.

I think the idea would deserve to be its own issue and explored independently. It could be a seen as a v2 of `nexus typegen` cli command.

**Comment by tgriesser** [comment]
> The overarching pattern that I think we should be heading toward is that makeSchema config can be driven in code or by env vars, with as seamless as possible name mapping.

Possibly. I actually was going to push back on the first env var being added, but I figured if we needed we could rip it out later once plugins are finalized. I didn't see the extra value it adds vs. just letting the user do it themselves, though I see the appeal for tools that are looking to build on top of Nexus so I get that the benefit may outweigh the additional internal config.

> But we still have a relatively high bar here requiring 1) use ts-node 2) use --transpile-only flag. I think these are arguably internal details the user shouldn't have to think about. So perhaps the solution is not about (or just about) a domain specific env var but rather a higher level cli command:

I disagree that it's a high bar. I think we should err toward just documenting these things and educating the library consumer, instead of trying to make things work automagically and adding more maintenance on our end. I tend to feel the same about what we're doing with the auto-gen into `@types` that it'd be better to educate/explain what's going on rather than try to make it "just work", causing more confusion when it "just doesn't".

> To make things simpler we could make entrypoint arg optional, default being top-down search short-circuiting on first match:

I think this would be fine for something like yoga or nexus-prisma to do, but I'd prefer to not make this a part of Nexus. That pattern would not match any of the file conventions I have used personally, so I'd rather just tell folks to be explicit since they'll know what file they've defined their schema in.

> when would you want NEXUS_SHOULD_GENERATE_ARTIFACTS but not NEXUS_EXIT_AFTER_TYPEGEN?

Yeah, if we're going to do an env var, I think we should consolidate into a single `NEXUS_TYPEGEN` which would generate, exit, and do some light console.logging.

> Had a nice conversation with @timsuchanek and he raised an interesting idea of exploring the potential of performing Nexus typegen with with TypeScript Plugins. It seems there could be some nice benefits to this.

Sounds interesting, would like learn more about the possible benefits here.

**Comment by jasonkuhrt** [comment]
The approach taken since opening the PR evolved. It is now based on a `NEXUS_CONFIG` env var allowing JSON to be passed down that is merged (and overrides) makeSchema code config (minus types, plugins).

Orthogonal to CLI.

`NEXUS_CONFIG` looks to me to be an easier interface for internals, users, and machines (e.g. cli). While it's more verbose for users, it's trivial to use once one understands that any documented config can be passed as JSON.

**Comment by tgriesser** [comment]
This will be fixed by #242

**Comment by tgriesser** [comment]
Added in the latest 0.12.0-rc.4 

Docs: https://nexus.js.org/docs/api-makeschema#shouldexitaftergenerateartifacts
